### ISSUE TITLE
Recipe: safeguard script to allow manipulate the web app in difficult situations

### ISSUE THREAD

**Issue by delapuente** [mozilla/serviceworker-cookbook#29]
From time to time, playing with SW and offline caches you end with a cached version of a bugged script controlling a critical part of your application. Designing the SW to always fetch a safeguard url and loading this script at the beginning of your web application prevent these kind of situation by allowing you to further control the application.

**Comment by delapuente** [comment]
After an update, offliner nevers activate the new version unless the client instruct it to do it. I was debugging why integration was not working and I realized I was not activating the new version in the client code but as this code was cached it never gets updated. With this kind of safeguard / backdoor script I could be able to inject an activation instruction. This is something I'm including in the current integration but I think it's not only a good but advisable too idea. What do you think @digitarald @mykmelez ?

**Comment by mykmelez** [comment]
> After an update, offliner nevers activate the new version unless the client instruct it to do it. I was debugging why integration was not working and I realized I was not activating the new version in the client code but as this code was cached it never gets updated.

Perhaps I don't understand the feature or its intention well enough, but it seems dangerous for *offliner* not to activate an updated service worker by default on reload (or navigate). Partly that's because of the problem you've identified here, but mostly it's because "activate on reload" is fundamental to the service worker update model. All the resources I've read about the update model describe this behavior, including [Introduction to Service Worker](http://www.html5rocks.com/en/tutorials/service-worker/introduction/#toc-how) and [ServiceWorkers Explained](https://github.com/slightlyoff/ServiceWorker/blob/master/explainer.md).

Some of those resources do describe ways to activate sooner than reload, including your infographic. But none of them describe ways to delay activation past reload. It may nevertheless be useful to do so in some cases, but it's a footgun to make that the default behavior, since it's too easy for a developer with some knowledge of service workers to assume that their updated service worker will be activated automatically in that case, and neglect to activate it explicitly. Even you, the developer of *offliner*, fell into that trap!

So I would not make that the default behavior but rather make *offliner* activate an updated service worker by default on reload. Then, I would make it an option to delay activation past (re)load if there's a use case for that, but I would warn developers about its dangers in the documentation!

> With this kind of safeguard / backdoor script I could be able to inject an activation instruction. This is something I'm including in the current integration but I think it's not only a good but advisable too idea. What do you think @digitarald @mykmelez ?

Isn't it possible to circumvent the current active service worker via a force refresh, i.e. shift+reload? It seems like this is the kind of workaround that would be better to have in developer tools, given that the use case is to recover from errors during development, rather than to have it in code that gets shipped to all users.

**Comment by delapuente** [comment]
I see your point @mykmelez but we are not talking about the same concept of activation. My bad for not introducing some context.

The worker gets activated on reload, of course. Actually because it uses _skip & claim_. What is not activated in the sense of the app lifecycle is the new version. Upon an update (and this is not a SW update but the strategy plugged inside offliner) _offliner_ downloads the new version and populate a separated cache but it does not start serving from there. When you use `offliner.activate()` on the offliner client you order _offliner_ to start using the new cache.

@mykmelez does it make more sense to you?

**Comment by mykmelez** [comment]
> @mykmelez does it make more sense to you?

It does, thanks for the context! The piece I guess I'm still missing is *why* the *offliner* implementation doesn't start using the new cache by default when the updated service worker is activated.

I understand that there can be some value to making this configurable, such that a client should be able to activate the updated service worker without starting to use the new cache (although I would still be interested to hear an explanation of the use case for this feature, as it isn't immediately clear to me).

But I don't understand the value in making this the *default* behavior. Wouldn't it make more sense for *offliner* to start using the new cache immediately unless the client has configured it to delay using the cache until an explicit call to *activate*?

Again, it's possible that I'm misunderstanding the common use cases for *offliner*. I'm imagining that *offliner* is useful for a developer who prefers to write their own service worker—and have complete control over its lifecycle—over using a high-level tool like Oghliner that writes it for them; but at the same time is open to using third-party tools to make writing and maintaining that service worker easier (i.e. they don't insist on writing it completely from scratch and maintaining it entirely by themselves).

As that developer, I would expect *offliner* to define a set of default behaviors while giving me the ability to configure it to behave differently. And regarding the question of when to start using a new cache, I would expect the default behavior to be that *offliner* starts using the new cache as soon as it activates the new worker, since that would be the common case, unless I configure it to delay using the new cache until I explicitly call *activate*.

Do I misunderstand the common case, or am I perhaps not actually thinking like the developers you're targeting with *offliner*?

**Comment by delapuente** [comment]
I think it is better to combine the discussion with mozilla/wadi#14 in order to not introduce noise in the recipe.

**Comment by mykmelez** [comment]
> I did not say you misunderstood the common case but we were thinking about different common cases. I think it is better to combine the discussion with mozilla/wadi#14 in order to not introduce noise in the recipe. See my comments there.

Ok, although I actually think the conversation belongs in this thread. The point I'm trying to make is that service workers (and libraries that help you to implement service workers, like *offliner*) should not be designed with footguns that you have to protect yourself against with foot armor like this recipe.

Rather than write recipes that show people how to arm their feet, I would write recipes that show people how to not shoot themselves in the foot.

**Comment by delapuente** [comment]
> The point I'm trying to make is that service workers (and libraries that help you to implement service workers, like offliner) should not be designed with footguns that you have to protect yourself against with foot armor like this recipe.

You're right but service workers in addition to offline caches have the intrinsic problem they could cache content you don't intended to cache. What we can do is to encourage good practices like the one in this recipe. If the developer get used to implement this recipe, he always could add a piece of JS (even in production environments) to solve the potential mess. 

> Rather than write recipes that show people how to arm their feet, I would write recipes that show people how to not shoot themselves in the foot.

This recipe is not intended to arm your foot, but to encourage you to have the medical kit near, just in case.

Anyway, if you consider it is not worth the effort or even worse, it is harmful or an antipattern, feel free to close it.

**Comment by mykmelez** [comment]
> Anyway, if you consider it is not worth the effort or even worse, it is harmful or an antipattern, feel free to close it.

I'm not the technical lead for this project, so I won't make that decision. But I do think it would be more valuable to create a recipe that shows how to design a service worker that doesn't run the risk of caching something that prevents the developer from updating a part of their app (whether in a development or production environment).

**Comment by mykmelez** [comment]
> I'm not the technical lead for this project, so I won't make that decision. But I do think it would be more valuable to create a recipe that shows how to design a service worker that doesn't run the risk of caching something that prevents the developer from updating a part of their app (whether in a development or production environment).

On the other hand, Flipkart apparently [found a feature like this useful](http://tech-blog.flipkart.net/2015/11/progressive-web-app/): "One such significant challenge that emerged from our use of Service Workers was to build a “kill switch”. It is easy to end up with bugs in Service Workers and stale responses. Having a reliable mechanism to purge all caches has helped us to be proactively ready for any contingencies or surprises."

**Comment by digitarald** [comment]
To get to some code that we can iterate on, maybe we should just call the recipe kill-switch and solve the simplified cache-nuking use case.

Would that solve your user story from comment 2, @delapuente?

**Comment by delapuente** [comment]
Totally, it is the same case indeed. I ended up with a bugged version of my update script so it never updated the worker. :) You can completely mess with the worker in such a way your kill switch will be completely useless but let's try to educate the developer to anticipate and prevent this.

**Comment by delapuente** [comment]
> But I do think it would be more valuable to create a recipe that shows how to design a service worker that doesn't run the risk of caching...

That's the point, the recipe would try to illustrate a safe worker, this safe worker allow to press the _kill switch_ when needed.

**Comment by mykmelez** [comment]
> That's the point, the recipe would try to illustrate a safe worker, this safe worker allow to press the kill switch when needed.

Erm, I actually meant something slightly different: a recipe that demonstrates a service worker that will never get into a state that would require a kill switch to get out of. I.e. a recipe for which it would never be necessary to have a kill switch.

It may also be useful to demonstrate a service worker that contains a kill switch. But I think of those as two different recipes. The first one illustrates design patterns that make a kill switch unnecessary. The second one illustrates the design an implementation of a kill switch that you can choose to implement *instead of* designing a service worker that makes a kill switch unnecessary.

**Comment by delapuente** [comment]
But the risk of making such a worker is never completely unavoidable as the API allows you to do it. Practically all the recipes of _more than offline_ don't fall into that pitfall because they don't cache. But if you're caching, you eventually can fall in this trap. Anyway, I'll try to design an SW pattern that never gets into this state.

Feedback is welcome. How do you imagine we can perform a complete cache of a site but ensuring you don't cache bugged code?

**Comment by mykmelez** [comment]
> Feedback is welcome. How do you imagine we can perform a complete cache of a site but ensuring you don't cache bugged code?

I'm unsure exactly what this would look like. Generally, I would design the service worker to ensure that cache entries can always be invalidated, regardless of whether or not their code is buggy. What that means specifically may depend on the details of the situation, however.

Do you have an example of a service worker that caches buggy code that prevents the service worker from updating that code?
